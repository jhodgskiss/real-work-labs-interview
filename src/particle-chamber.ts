class Particle {
    position: number;
    speed: number;

    constructor(position: number, speed: number) {
        this.position = position;
        this.speed = speed;
    }

    // advance particle and return whether it is still in the chamber
    step(chamberLength: number): boolean {
        this.position += this.speed;
        return this.speed > 0 ? this.position < chamberLength : this.position >= 0;
    }
}

class ParticleChamber {
    particles: Particle[];
    length: number;

    constructor(initialPosition: string, speed: number) {
        this.particles = [];
        this.length = initialPosition.length;

        Array.from(initialPosition).forEach((char, index) => {
            switch (char) {
            case 'R':
                this.addParticle(index, speed);
                break;
            case 'L':
                this.addParticle(index, -speed);
                break;
            default:
                break;
            }
        });
    }

    addParticle(position: number, speed: number) {
        this.particles.push(new Particle(position, speed));
    }

    getChamberState(): string {
        const chamberArray = Array(this.length).fill('.');
        this.particles.forEach(particle => chamberArray[particle.position] = 'X');
        return chamberArray.join('');
    }

    // advances particles and removes those outside chamber
    step() {
        this.particles = this.particles.filter(particle => particle.step(this.length));
    }
}

export function animate(initialPosition: string, speed: number) {
    if (speed === 0) {
        throw new Error('A speed of zero would result in an infinite loop, check inputs.');
    }

    if (!Number.isInteger(speed)) {
        throw new Error('Floats currently not supported.');
    }

    const chamber = new ParticleChamber(initialPosition, speed);
    const chamberArrays: string[] = [chamber.getChamberState()];

    while (chamber.particles.length) {
        chamber.step();
        chamberArrays.push(chamber.getChamberState());
    }

    return chamberArrays;
}
