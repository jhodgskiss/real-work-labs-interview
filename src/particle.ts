class Particle {
    position: number;
    speed: number;

    constructor(position: number, speed: number) {
        this.position = position;
        this.speed = speed;
    }

    getDistanceToExit(chamberLength: number): number {
        return this.speed > 0 ? chamberLength - this.position : this.position;
    }

    step(chamberLength: number): boolean {
        this.position += this.speed;
        return this.speed > 0 ? this.position < chamberLength : this.position >= 0;
    }
}

export function animate(initialPosition: string, speed: number) {
    const particles: Particle[] = [];
    const chamberLength: number = initialPosition.length;

    if (speed === 0) {
        throw new Error('A speed of zero would result in an infinite loop, check inputs.');
    }

    if (!Number.isInteger(speed)) {
        throw new Error('Floats currently not supported.');
    }

    Array.from(initialPosition).forEach((char, index) => {
        switch (char) {
        case 'R':
            particles.push(new Particle(index, speed > 0 ? speed : -speed));
            break;
        case 'L':
            particles.push(new Particle(index, speed > 0 ? -speed : speed));
            break;
        default:
            break;
        }
    });

    const maxDistance = Math.max(...particles.map(particle => particle.getDistanceToExit(chamberLength)), 0);
    const maxSteps = Math.abs(Math.ceil(maxDistance / speed));

    const particleArrays: string[][] = [...Array(maxSteps + 1)].map(() => Array(chamberLength).fill('.'));

    particles.forEach(particle => {
        let time = 0;
        do {
            particleArrays[time++][particle.position] = 'X';
        } while (particle.step(chamberLength));
    });

    return particleArrays.map(charArray => charArray.join(''));
}
