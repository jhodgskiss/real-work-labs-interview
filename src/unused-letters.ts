const _CHARACTER_SET = 'abcdefghijklmnopqrstuvwxyz';
const _CHARACTERS = new Set(Array.from(_CHARACTER_SET));

export function getUnusedCharacters(string: string): string {
    const usedCharacterSet = new Set(Array.from(string.toLowerCase()));
    const unusedCharacters = _CHARACTERS.difference(usedCharacterSet);
    return Array.from(unusedCharacters).join('');
}
