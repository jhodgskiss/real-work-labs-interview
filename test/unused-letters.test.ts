import { getUnusedCharacters } from "../src/unused-letters";

test('example 0',() => {
    expect(getUnusedCharacters("A slow yellow fox crawls under the proactive dog")).toMatchSnapshot();
});

test('example 1', () => {
    expect(getUnusedCharacters("A quick brown fox jumps over the lazy dog")).toMatchSnapshot();
});

test('empty string', () => {
    const alphabet = Array(26)
        .fill('')
        .map((_,index) => String.fromCharCode('a'.charCodeAt(0) + index)).join('');
    expect(getUnusedCharacters('')).toMatch(alphabet); // sanity check
    expect(getUnusedCharacters('')).toMatchSnapshot();
});

test('capital letters', () => {
    expect(getUnusedCharacters("ABCDEF")).toMatchSnapshot();
});
