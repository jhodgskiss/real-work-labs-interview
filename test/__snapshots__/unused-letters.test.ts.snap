// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`capital letters 1`] = `"ghijklmnopqrstuvwxyz"`;

exports[`empty string 1`] = `"abcdefghijklmnopqrstuvwxyz"`;

exports[`example 0 1`] = `"bjkmqz"`;

exports[`example 1 1`] = `""`;
